application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/AutomaticDependenciesUpgradeJob.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.licensing.internal.upgrades;
+
+import org.quartz.Job;
+import org.quartz.JobExecutionContext;
+import org.quartz.JobExecutionException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.xpn.xwiki.plugin.scheduler.AbstractJob;
+import com.xpn.xwiki.web.Utils;
+
+/**
+ * Scheduler job that upgrades to the last compatible version the extensions that have a license.
+ * 
+ * @since 1.17
+ * @version $Id$
+ */
+public class AutomaticDependenciesUpgradeJob extends AbstractJob implements Job
```

Comment:
Would be better to have `LicensedExtentionsUpgradeJob`, since this job starts the upgrade for one or multiple extensions? Or the singular version is in general preferred?

Reply:
I think the singular form is more common. It's somehow implied that if a component can handle some entity then it should be able to handle multiple entities of the same type. In this case, if the job knows how to upgrade a single licensed extension then upgrading multiple licensed extensions is just a matter of iterating a list. Whether that list is received as input or computed is a detail.